Support GA loading asyncly

benmccallum opened the pull request:
PR for issue #43.

Zizzamia commented:
A couple of quick note:
1) remove `package-lock.json` from the PR, and don't worry to change the Perfume version. I will take care of it.
2) I'm not sure why it seems this file got renamed
```
__tests__/emnulated-performance.spec.ts → __tests__/emulated-performance.spec.ts
```

benmccallum commented:
1. Will do.
2. I fixed your spelling mistake in the filename 😉

Zizzamia commented:
@benmccallum Oooo, now I see the typo! Lol, thank you :)

benmccallum commented:
I've basically finished this but need to fix up the unit tests. 

Ended up with an `IAnalyticsTracker` implemented by a `GoogleAnalyticsTracker`, but the GA tracker is internal to perfume, not something that others can pass in themselves as I think you were worried about. I think it will work well. 

Do you think there would ever be a need for a consumer to pass in multiple custom IAnalyticsTrackers via the config? Right now the options only accepts one, but internally I store the instances in an array so it's easy enough to accept an array in the options too if we ever need to down the track. Can't see it being used though... you?

Zizzamia commented:
Let's store only one tracker for each **Perfume** instance. In case someone needs multiple trackers, she/he can create multiple Perfume instances.

benmccallum commented:
Ok, so it should be up to date with master now. All tests passing w/ 100% code coverage. I still haven't had time to test it in a browser though. Do you want to give it a go?

Couple of notes:
* Also, I renamed analyticsLogger to analyticsTracker, it's a breaking change, but I guess that would've been the case anyway since it's now an IAnalyticsTracker. So maybe it's a good thing that it's a new prop name as it signifies that things have changed, including type. Looks like you'll have to go to 1.0 if you want to follow semver.
* I've got the new classes/interfaces I added in separate files, but I didn't know what to called the file for IAnalyticsTracker as your filenames are lowercase with dashes. would it have been i-analytics-tracker.ts? I see you've got a performance-impl.ts, with IPerformance, but that doesn't make sense to me as usually *Impl is short for *Implementation, not the interface.

Do you want to do a build and try it in a browser? I've gotta get on to some other things.

Zizzamia commented:
Amazing, I will review it and test it more during this week 🤘
Great work, thank you for the PR
![happy-face](https://user-images.githubusercontent.com/829214/44754875-ac967700-aad8-11e8-8ae1-8957b1f830e8.gif)

Zizzamia commented on src/google-analytics-tracker.ts:
```diff
@@ -0,0 +1,37 @@
+  metricQueue: Metric[] = [];
+  timingVar: string = 'name';
+
+  canSend(): boolean {
```
For `boolean` type we already use the prefix convention `is`, can you name this method `isReady`.
Because is a `read-only` variable, we can use a Typescript Getter instead.
```
get canSend(): boolean {
   return window.ga && typeof window.ga === 'function';
}
```

benmccallum replied:
This will be removed anyway in upcoming simplification refactor.

Zizzamia commented on src/metric.ts:
```diff
@@ -0,0 +1,9 @@
+export default class Metric {
```
A better name for this class is **PerformanceEntry**, because we are storing the `name` and `duration` coming from the [Performance Observer](https://w3c.github.io/performance-timeline/#dom-performanceentry) and [Performance Mark](https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark) APIs.

benmccallum replied:
Done.

benmccallum replied:
Hm, there's a problem with this. It conflicts with an existing PerformanceEntry type in lib.dom.ts. I've namespaced it for now. Got another suggested name or is namespacing fine?

Zizzamia commented on src/metric.ts:
```diff
@@ -0,0 +1,9 @@
+  name: string;
+  duration: number;
+
+  constructor(metricName: string, duration: number) {
```
You can make it simpler and use `name` instead of `metricName`

benmccallum replied:
Done.

Zizzamia commented on src/perfume.ts:
```diff
@@ -174,42 +188,51 @@ export default class Perfume {
+   * window.load handler.
+   * Tries to send any queued metrics to trackers.
+   */
+  onWindowLoad(): void {
```
We use the prefix `did` for listener callback from the `document` or `window`, can you please change this to `didWindowLoad`.

benmccallum replied:
Done.

benmccallum commented on .gitignore:
```diff
@@ -33,3 +33,10 @@ testem.log
 .DS_Store
 Thumbs.db
+
+# VS Code - https://github.com/github/gitignore/blob/master/Global/VisualStudioCode.gitignore
```
Installed the editorconfig extension for VS Code and that gave me the desired result I wanted. I still think the VS Code gitignore lines are valid for those of us who use it (e.g. we could create an extensions.json and recommend the editorconfig extension is installed by devs), but for the sake of just getting this across the line I'll remove them.

benmccallum commented on src/perfume.ts:
```diff
@@ -70,6 +67,7 @@ export default class Perfume {
   observeFirstInputDelay: Promise<number>;
   observeTimeToInteractive: Promise<number>;
   timeToInteractiveDuration: number = 0;
+  analyticsTrackers: IAnalyticsTracker[] = [];
```
I liked the idea of there being an interface as people could come up with implementations for different providers and there'd be a mixture you could plug in as you pleased (i.e. documented implementations for different trackers). 

Removing this though to get it through.

benmccallum commented:
Hmm, I've made all the changes but really don't want to have to write unit tests for all this until you've approved it as I'm just wasting my life here...

benmccallum commented:
Hey @Zizzamia , do you think you could take a look at this? Keen to get it into master and onto npm so can add it to our site.

Zizzamia commented:
I'm not entirely convinced this is the best way to solve this problem, what's the current solution to make it work GTag with Perfume? Also, if you want, we can jump on a google hangout to discuss more of it.

I would love leverage more Promises with Async/Await to get the data out of
```js
perfume.firstPaintDuration;
perfume.firstContentfulPaintDuration;
perfume.firstInputDelayDuration;
```

benmccallum commented:
🤦‍♂️ . 

Why don't you just:
1. Rip out the analytics tracker/send part of this library entirely.
2. Then either:
2.1 Fire events on metric collected for us to listen to; and/or,
2.2 Expose option/s for us to add custom functions that are run on perfume metric collected.
3. Give a GA code snippet for sending the data to your own GA account or build a little npm package that does it and depends on perfume.js.

That way perfume.js stays concerned with recording metric data and the rest of us can build whatever we want on top of it.
